### PR TITLE
Cannot create full datetime product attribute 4053

### PR DESCRIPTION
**Pull request by novikor**
### Description (*)
  -  Added time picker for product edit page (for attrobutes with `frontend_type` = `date`)

### Fixed Issues (if relevant)

1. magento/magento2#4053: Cannot create full datetime product attribute

### Manual testing scenarios (*)
1. Go to product edit page
2. Find any `date` attribute
3. Try to save date & time value
4. Make sure value is saved correctly
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @novikor. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php [hunk 1]
```diff
@@ -0,0 +1,132 @@
+<?php
```
Comment (sivaschenko) [inline app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php]: New added classes should have strict mode enabled, also, can you please cover this class with unit test

File: app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                true
+            );
+            $dateObj->setTime($hour, $minute, $second);
+            //convert store date to default date in UTC timezone without DST
+            if ($setUtcTimeZone) {
+                $dateObj->setTimezone(new \DateTimeZone('UTC'));
+            }
+            return $dateObj;
+        } catch (\Exception $e) {
+            return null;
```
Comment (sivaschenko) [inline app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php]: I think it's better to log the exception message

**Review by sivaschenko — DISMISSED** [review]
@novikor Thank you for the contribution! Please see the review notes.

File: app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php (hunk 2 above)
Comment (novikor) [reply app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php]: @sivaschenko, Please recheck

File: app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php (hunk 1 above)
Comment (novikor) [reply app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php]: @sivaschenko strict types have been added, - it is my fault, just copied class and forgot to refactor it.
I agree that Unit tests are a good thing, but, in fact, there is nothing to cover.

**Comment by novikor** [comment]
@sivaschenko , any updates?

**Comment by sivaschenko** [comment]
@novikor thanks for your updates, I need a bit more time to process this pull request. One thing that already araised - these changes should be definitely covered by integration tests. Can you please provide the coverage?

**Comment by novikor** [comment]
@sivaschenko , will be done

**Comment by novikor** [comment]
@sivaschenko , please check. I found only one place when attribute input types are being covered by integrational tests and covered it in the same way.

Also, it seems that functional tests failed because of some troubles with travis/licenses.

**Comment by VladimirZaets** [comment]
Hi @novikor. Thanks for collaboration. Please explain why you need separate component for the datetime input if it is only one property of configuration option for date component? As for me, it's redundant, each of components have a lot of configuration options and creating separate components for each of them as for me is overhead.

**Comment by novikor** [comment]
@VladimirZaets , I thought it was a good idea to leave existing component the same to avoid any effect on existing usage of it in case of Magento upgrade on any project.

If it does not have any sense, I will remake it these week. Please let me know.

Thank you for review.

**Comment by novikor** [comment]
@VladimirZaets, I suggest also to mark the existing date component as deprecated and remove it in 2.4. So, only datetime component will be used in future.

**Comment by novikor** [comment]
Hi, @VladimirZaets . I did it.

**Comment by VladimirZaets** [comment]
Hi @novikor. Great, thank you!

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-4117](https://jira.corp.magento.com/browse/ENGCOM-4117) has been created to process this Pull Request

**Comment by VladimirZaets** [comment]
Hi @novikor. 

Your fix provide the next problem (manual testing scenario):
Create product "Test"
Click Edit for "Test" product
Set date for Set Product as New From
 Date is saved in format mm/dd/yyyy without time
Edit file app/code/Magento/Ui/view/base/web/js/lib/knockout/bindings/datepicker.js
Set showsTime: true
Clear pub/static and flush cache
Create product "Test1"
Click Edit for "Test1" product
Set date and time for Set Product as New From
 Date is saved in format mm/dd/yyyy 12:00 AM (time always is shown as 12:00 AM )

**Comment by novikor** [comment]
@VladimirZaets , thanks for the update.
I will check it in the nearest future.

**Comment by novikor** [comment]
@VladimirZaets , fixed.
I can`t imagine how could it work before.

**Comment by VladimirZaets** [comment]
@novikor thanks

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-4117](https://jira.corp.magento.com/browse/ENGCOM-4117) has been created to process this Pull Request

**Comment by VasylShvorak** [comment]
Testing result: Inconsistency between time in admin and time in database.
**Manual testing scenario:**

- Install magento with timezone `--timezone=America/Chicago` 
- Go to Catalog->Products
- Click Add Product
-- Fill required fileds
-- Set Product as New From = Go Today
- Save Product
- Compare time in admin and database
-  Time on product edit page is shown according timezone, time in database is shown according UTC(in my case)
#19385Admin
#19385DB


@sivaschenko @VladimirZaets Is this correct behavior?

**Comment by sivaschenko** [comment]
@VasylShvorak that's correct as long as the time in UI is displayed consistently.

**Comment by VasylShvorak** [comment]
Hi @novikor Time is displayed inconsistently. Time is saved,for example, in timezone=America/Chicago by Calendar widget and time is displayed in UTC on Products grid page.
 #19385ProdEdit 
 #19385ProdGrid

**Comment by novikor** [comment]
I am working on it.

**Comment by novikor** [comment]
Now correct timezone is set for `date` columns of catalog grid.
An extension point has been provided for that purpose.

Please review.

File: app/code/Magento/Catalog/Ui/Component/Column/DateConfigProvider.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    ) {
+        $this->scopeConfig = $scopeConfig;
+    }
+
+    /**
+     * @param string $dataType
+     *
+     * @return array
+     */
+    public function getConfig(string $dataType):array
```
Comment (sivaschenko) [inline app/code/Magento/Catalog/Ui/Component/Column/DateConfigProvider.php]: What is the purpose of `$dataType` argument for the concrete provider?

File: app/code/Magento/Catalog/Ui/Component/ColumnFactory.php [hunk 4]
```diff
@@ -43,11 +45,21 @@ class ColumnFactory
...
-    {
+    private $configProvider;
+
+    /**
+     * @param \Magento\Framework\View\Element\UiComponentFactory                   $componentFactory
+     * @param \Magento\Catalog\Ui\Component\Column\DataTypeConfigProviderInterface $configProvider
+     */
+    public function __construct(
+        \Magento\Framework\View\Element\UiComponentFactory $componentFactory,
+        DataTypeConfigProviderInterface $configProvider
```
Comment (sivaschenko) [inline app/code/Magento/Catalog/Ui/Component/ColumnFactory.php]:
```suggestion
        DataTypeConfigProviderInterface $configProvider = null
```

**Review by sivaschenko — DISMISSED** [review]
Hi @novikor thanks for the fix and the extension point. Please see my review comments

File: app/code/Magento/Catalog/Ui/Component/Column/DateConfigProvider.php (hunk 3 above)
Comment (novikor) [reply app/code/Magento/Catalog/Ui/Component/Column/DateConfigProvider.php]: Actually, there is no purpose. However, they should implement the same interface.
What about changing it to `string $dataType = null`?

Comment (sivaschenko) [reply app/code/Magento/Catalog/Ui/Component/Column/DateConfigProvider.php]: @novikor I don't think the pool and the provider should implement the same interface in this case (as pool's interface does not make sense for the concrete provider)

Comment (novikor) [reply app/code/Magento/Catalog/Ui/Component/Column/DateConfigProvider.php]: @sivaschenko , please recheck my solution.

**Comment by VasylShvorak** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @VasylShvorak. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @VasylShvorak, here is your [new Magento instance](https://pr-19385.instances.magento-community.engineering).
Admin access: [https://pr-19385.instances.magento-community.engineering/admin](https://pr-19385.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by VasylShvorak** [comment]
Hi @novikor during testing PR changes I got an error: `Class Magento\Catalog\Ui\Component\DataTypeConfigProviderInterface does not exist` when navigate to Catalog->Products.
#19385Exc

Could you take a look?

**Comment by novikor** [comment]
@VasylShvorak , please recheck.
It was a silly mistake from my side.

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-4117](https://jira.corp.magento.com/browse/ENGCOM-4117) has been created to process this Pull Request

**Comment by VasylShvorak** [comment]
:heavy_check_mark: QA passed

**Comment by p-bystritsky** [comment]
Hello, @novikor. Please fix the test failures listed below.

**Comment by sidolov** [comment]
@novikor , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by novikor** [comment]
**Notice**: please recheck https://github.com/magento/magento2/issues/23113

**Comment by engcom-Delta** [comment]
Hi @novikor I cannot reproduce issue [#23113](https://github.com/magento/magento2/issues/23113) on 2.3-develop.
English locale:
#19385English

French locale:
#19385French

**Comment by sidolov** [comment]
Hi @novikor , since mentioned issue not reproducing I'm closing this PR. Feel free to reopen if you find one more case. Thank you for your contribution!

**Pull request closed without merging**